### PR TITLE
M2P-240 Add support for module Amasty Reward Points

### PR DESCRIPTION
**Pull request by dumega**
# Description
Amasty Reward Points - General plugin support

Fixes: https://boltpay.atlassian.net/browse/M2P-240

#changelog Add support for module Amasty Reward Points

# Type of change

- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please validate that you have tested your change in at least one of the following areas:

- [ ] Successfully tested locally (or docker image)
- [x] Successfully tested on a staging or sandbox server
- [ ] Successfully tested on a merchant's staging server

# For PR Reviewer 
- [ ] Reviewed unit tests to make sure we are using real components rather than mocks as much as possible?
- [ ] For any major change (observer, new Bolt feature, core Magento interaction) we must add a feature switch, did you verify this?

# Checklist:

- [x] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have created or modified unit tests to sufficiently cover my changes.
- [ ] I have added my Jira ticket link and provided a changelog message.

### CODE REVIEW

**Review by danac-gs — APPROVED** [review]

File: ThirdPartyModules/Amasty/Rewards.php
Comment (johnmccombs1) [inline ThirdPartyModules/Amasty/Rewards.php:58]: @dumega I added the parent quote as a parameter in for the filter collectDiscounts in this PR https://github.com/BoltApp/bolt-magento2/pull/931/files#diff-d9d65ba4532ae42767de8b35f1843726R2178
Can you re-base the branch to master and add it as well?

**Review by johnmccombs1 — CHANGES_REQUESTED** [review]

**Review by matt-thomason — DISMISSED** [review]
LGTM, pending re-review from @johnmccombs1

File: ThirdPartyModules/Amasty/Rewards.php
Comment (dumega) [reply ThirdPartyModules/Amasty/Rewards.php:59]: fixed

File: ThirdPartyModules/Amasty/Rewards.php
Comment (danac-gs) [inline ThirdPartyModules/Amasty/Rewards.php:74]: @dumega can you explain the logic in this loop a bit better?
Why do we need to recreate the whole discount array instead of just adding to it?
Seems like we can potentially lose the internal coupon discount here, if applied, but I may be wrong.
Did you test with both Magento coupon and Amasty Reward applied?

Comment (dumega) [reply ThirdPartyModules/Amasty/Rewards.php:74]: @danac-gs Actually that's the purpose of this PR, let me show you some screenshots

image
image

As you can see, the Amasty reward points combine with Magento coupon, and when the Bolt plugin try to collect discount amount from address object, it just get the reward points and Magento coupon as a whole, and then sent to the Bolt server as coupon discount. So this PR is to separate Amasty reward points from Magento coupon.

image

Comment (danac-gs) [reply ThirdPartyModules/Amasty/Rewards.php:74]: Great, thanks.

**Review by danac-gs — APPROVED** [review]

**Review by vitaliyreznikov — APPROVED** [review]

**Review by johnmccombs1 — APPROVED** [review]
LGTM

**Pull request merged**
